<a href="https://colab.research.google.com/github/h586601/ML2_DAT158/blob/main/ML2_Group7_BoxOfficePrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

First we must upload training data. We will upload test data at the same time just because its simpler

In [ ]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from google.colab import drive 
drive.mount('/content/drive')

train = pd.read_csv('drive/MyDrive/ColabNotebooks/train.csv')
test = pd.read_csv('drive/MyDrive/ColabNotebooks/test.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     3000 non-null   int64  
 1   belongs_to_collection  604 non-null    object 
 2   budget                 3000 non-null   int64  
 3   genres                 2993 non-null   object 
 4   homepage               946 non-null    object 
 5   imdb_id                3000 non-null   object 
 6   original_language      3000 non-null   object 
 7   original_title         3000 non-null   object 
 8   overview               2992 non-null   object 
 9   popularity             3000 non-null   float64
 10  poster_path            2999 non-null   object 
 11  production_companies   2844 non-null   object 
 12  production_countries   2945 non-null   object 
 13  release_date           3000 non-null   object 
 14  runtime                2998 non-null   float64
 15  spok

In [ ]:
train.head()

id  ...   revenue
0   1  ...  12314651
1   2  ...  95149435
2   3  ...  13092000
3   4  ...  16000000
4   5  ...   3923970

[5 rows x 23 columns]

In [ ]:
explore_train = train.copy()

In [ ]:
# Delete some unneccessary columns for a better overview
explore_train = explore_train.drop(["id", "homepage", "imdb_id", "overview", "poster_path", "tagline", "title"], axis=1)

In [ ]:
# Fill in NaN values with 0 and other values with 1
def NaNToBinary(feature_name):
    explore_train[feature_name].fillna(0, inplace=True)
    
    array = explore_train[feature_name].copy()
    
    for i in range(len(array)):
        if array[i] != 0:
            array[i] = 1
        
    return array

In [ ]:
belong = "belongs_to_collection"
belongs_to_collection = NaNToBinary(belong)
explore_train = explore_train.drop(belong, axis=1)

In [ ]:
pd.DataFrame(belongs_to_collection.value_counts())

belongs_to_collection
0                   2396
1                    604

In [ ]:
explore_train.head()

budget  ...   revenue
0  14000000  ...  12314651
1  40000000  ...  95149435
2   3300000  ...  13092000
3   1200000  ...  16000000
4         0  ...   3923970

[5 rows x 15 columns]

In [ ]:
explore_train["belongs_to_collection"] = belongs_to_collection

In [ ]:
explore_train["status"].value_counts()

Released    2996
Rumored        4
Name: status, dtype: int64

In [ ]:
rumor = explore_train[explore_train.status == "Rumored"]
rumor.head()

budget  ... belongs_to_collection
609        0  ...                     1
1007       0  ...                     0
1216       0  ...                     0
1618  891000  ...                     0

[4 rows x 16 columns]

In [ ]:
def categoryToBinary(feature_name, cat1, cat2):
    array = explore_train[feature_name].copy()
    
    for i in range(len(array)):
        if array[i] == cat1:
          array[i] = 1
        if array[i] == cat2:
          array[i] = 0
        
    return array

In [ ]:
# I want to keep the status feature, so I change it to binary values
status = categoryToBinary("status", "Released", "Rumored")

In [ ]:
pd.DataFrame(status.value_counts())

status
1    2996
0       4

In [ ]:
explore_train["status"] = status

In [ ]:
explore_train.head()

budget  ... belongs_to_collection
0  14000000  ...                     1
1  40000000  ...                     1
2   3300000  ...                     0
3   1200000  ...                     0
4         0  ...                     0

[5 rows x 16 columns]

In [ ]:
explore_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   budget                 3000 non-null   int64  
 1   genres                 2993 non-null   object 
 2   original_language      3000 non-null   object 
 3   original_title         3000 non-null   object 
 4   popularity             3000 non-null   float64
 5   production_companies   2844 non-null   object 
 6   production_countries   2945 non-null   object 
 7   release_date           3000 non-null   object 
 8   runtime                2998 non-null   float64
 9   spoken_languages       2980 non-null   object 
 10  status                 3000 non-null   object 
 11  Keywords               2724 non-null   object 
 12  cast                   2987 non-null   object 
 13  crew                   2984 non-null   object 
 14  revenue                3000 non-null   int64  
 15  belo

In [ ]:
explore_train['status'] = explore_train['status'].astype(int)
explore_train['belongs_to_collection'] = explore_train['belongs_to_collection'].astype(int)
explore_train['popularity'] = explore_train['popularity'].astype(int)

In [ ]:
explore_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   budget                 3000 non-null   int64  
 1   genres                 2993 non-null   object 
 2   original_language      3000 non-null   object 
 3   original_title         3000 non-null   object 
 4   popularity             3000 non-null   int64  
 5   production_companies   2844 non-null   object 
 6   production_countries   2945 non-null   object 
 7   release_date           3000 non-null   object 
 8   runtime                2998 non-null   float64
 9   spoken_languages       2980 non-null   object 
 10  status                 3000 non-null   int64  
 11  Keywords               2724 non-null   object 
 12  cast                   2987 non-null   object 
 13  crew                   2984 non-null   object 
 14  revenue                3000 non-null   int64  
 15  belo

In [ ]:
corr_matrix = explore_train.corr()
corr_matrix["revenue"].sort_values(ascending=False)

revenue                  1.000000
budget                   0.752965
popularity               0.461075
belongs_to_collection    0.339425
runtime                  0.216380
status                   0.016806
Name: revenue, dtype: float64

In [ ]:
TEXT_FIELDS =[ ("genres", "name"), 
                ("production_companies", "id"),                
                ("production_countries", "iso_3166_1"),
                ("spoken_languages", "name"),
                ("Keywords", "name"),
                ("cast", "name"),                         
                ("crew", "name"),                
             ]

SPLITTER = " "

In [ ]:
import ast
from sklearn.base import BaseEstimator, TransformerMixin

class InfoExtractor(BaseEstimator, TransformerMixin):
    def __init__(self, field, replace=False):
        self.field = field
        self.replace = replace

    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        clone_X = X.copy()   
        for feature, field_name in self.field:
            if self.replace:
                clone_X[feature] = X[feature].apply(lambda x: self.extract_field(x, field_name))
            else:
                clone_X["{0}_{1}".format(feature, field_name)] = X[feature].apply(lambda x: self.extract_field(x, field_name))
        return clone_X
        
    def extract_field(self, data, field_name):
        if(data is not np.nan):
            info = ast.literal_eval(data)            
            result = SPLITTER.join("{}".format(x[field_name]).replace(SPLITTER, "_") for x in info)
            return result
        return np.nan
    
infoExtractor = InfoExtractor(field=TEXT_FIELDS, replace=True)
result = infoExtractor.transform(explore_train)
#filters = list(result.filter(regex="collection|genres|crew").columns)
filters = list(result.filter(regex="collection|cast|production|genres|languages|Keywords|crew").columns)
result[filters].head()

genres  ... belongs_to_collection
0                       Comedy  ...                     1
1  Comedy Drama Family Romance  ...                     1
2                        Drama  ...                     0
3               Thriller Drama  ...                     0
4              Action Thriller  ...                     0

[5 rows x 8 columns]

# **Information**

So far we have changed NaN to 0 and the rest to 1 in the belongs_to_collection, so we know that either it belongs to a collection or not. And changed status from "Released" and "Rumored" to 1 and 0, respectively.

Then changed belongs_to_collection and status from object types to int, and ran it through the correlation again.

Quite a positive correlation between revenue and if a movie belongs to a collection, but we must remember that our preparation of the data might trigger a false correlation.

# **Setting up the environment**

In [ ]:
!pip install pycaret

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 3021, in _dep_map
    return self.__dep_map
  File "/usr/local/lib/python3.7/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 2815, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/cli/base_command.py", line 180, in _main
    status = self.run(options, args)
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/cli/req_command.py", line 199, in wrapper
    return func(self, options, args)
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/commands/install.py", line 319, in run
    reqs, check_supported_wheels=not options.target_dir
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/resolution/resolvelib/resolver.py", l

In [ ]:
#Followed these instructions https://github.com/pycaret/pycaret/blob/master/tutorials/Regression%20Tutorial%20Level%20Beginner%20-%20REG101.ipynb

from pycaret.regression import *

# For Colab
from pycaret.utils import enable_colab
enable_colab()

Colab mode enabled.


In [ ]:
exp_reg = setup(data = explore_train, target = 'revenue', session_id=123, numeric_features=['status','belongs_to_collection'])

IntProgress(value=0, description='Processing: ', max=3)

Initiated  . . . . . . . . . . . . . . . . . .            16:00:48
Status     . . . . . . . . . . . . . . . . . .  Preprocessing Data

Text(value="Following data types have been inferred automatically, if they are correct press enter to continue…

Data Type
budget                     Numeric
genres                 Categorical
original_language      Categorical
original_title         Categorical
popularity                 Numeric
production_companies   Categorical
production_countries   Categorical
release_date                  Date
runtime                    Numeric
spoken_languages       Categorical
status                     Numeric
Keywords               Categorical
cast                   Categorical
crew                   Categorical
belongs_to_collection      Numeric
revenue                      Label

The dataset has transformed from (3000, 16) to (2099, 1216), so the number of features have increased from 16 to 1216 because of the categories.

In [ ]:
best = compare_models()

In [ ]:
reg_model = create_model('rf')

In [ ]:
tuned_reg_model = tune_model(reg_model, n_iter=5, optimize = 'RMSLE')

In [ ]:
plot_model(tuned_reg_model)

In [ ]:
plot_model(reg_model)

In [ ]:
evaluate_model(tuned_reg_model)

In [ ]:
predict_model(tuned_reg_model)

In [ ]:
final_reg_model = finalize_model(tuned_reg_model)

In [ ]:
save_model(final_reg_model, 'BoxOfficeModel')

In [ ]:
saved_final_reg_model = load_model('BoxOfficeModel')